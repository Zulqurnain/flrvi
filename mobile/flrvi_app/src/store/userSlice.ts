import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';
import { RootState } from './store';

const API_URL = 'http://localhost:5000/api/v1';

export const createProfile = createAsyncThunk<
  any,
  { birthday: any; gender: string; photos: any[] },
  { rejectValue: { msg: string } }
>('user/createProfile', async (profileData, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.post(`${API_URL}/profile`, profileData, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const updateUserProfile = createAsyncThunk<
  any,
  { bio: string; photos: string[] },
  { rejectValue: { msg: string } }
>('user/updateProfile', async (profileData, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.put(`${API_URL}/profile`, profileData, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const registerUser = createAsyncThunk<
  { token: string },
  { name: string; email: string; password: string },
  { rejectValue: { msg: string } }
>('user/register', async (userData, { rejectWithValue }) => {
  try {
    const response = await axios.post(`${API_URL}/users/register`, userData);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const loginUser = createAsyncThunk<
  { token: string },
  { email: string; password: string },
  { rejectValue: { msg: string } }
>('user/login', async (credentials, { rejectWithValue }) => {
  try {
    const response = await axios.post(`${API_URL}/users/login`, credentials);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const fetchProfiles = createAsyncThunk<
  any[],
  void,
  { rejectValue: { msg: string } }
>('user/fetchProfiles', async (_, { rejectWithValue }) => {
  try {
    const response = await axios.get(`${API_URL}/profile/search`);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const fetchConversations = createAsyncThunk<
  any[],
  void,
  { rejectValue: { msg: string } }
>('user/fetchConversations', async (_, { rejectWithValue }) => {
  try {
    const response = await axios.get(`${API_URL}/chat/conversations`);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const fetchUserProfile = createAsyncThunk<
  any,
  string,
  { rejectValue: { msg: string } }
>('user/fetchUserProfile', async (userId, { rejectWithValue }) => {
  try {
    const response = await axios.get(`${API_URL}/profile/user/${userId}`);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const fetchCurrentUserProfile = createAsyncThunk<
  any,
  void,
  { rejectValue: { msg: string } }
>('user/fetchCurrentUserProfile', async (_, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.get(`${API_URL}/profile/me`, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const likeUserProfile = createAsyncThunk<
  any,
  string,
  { rejectValue: { msg: string } }
>('user/likeUserProfile', async (userId, { rejectWithValue }) => {
  try {
    const response = await axios.post(`${API_URL}/profile/like/${userId}`);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const subscribeUser = createAsyncThunk<
  any,
  { email: string; plan: string; source: string },
  { rejectValue: { msg: string } }
>('user/subscribe', async (subscriptionData, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.post(`${API_URL}/payment/subscribe`, subscriptionData, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const purchaseBoost = createAsyncThunk<
  any,
  { productId: string; amount: number },
  { rejectValue: { msg: string } }
>('user/purchaseBoost', async (purchaseData, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.post(`${API_URL}/payment/purchase-consumable`, purchaseData, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const fetchBlockedUsers = createAsyncThunk<
  any[],
  void,
  { rejectValue: { msg: string } }
>('user/fetchBlockedUsers', async (_, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.get(`${API_URL}/users/blocked`, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const unblockUser = createAsyncThunk<
  any,
  string,
  { rejectValue: { msg: string } }
>('user/unblockUser', async (userId, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.post(`${API_URL}/users/unblock`, { userId }, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const updateNotificationSetting = createAsyncThunk<
  any,
  { [key: string]: boolean },
  { rejectValue: { msg: string } }
>('user/updateNotificationSetting', async (settings, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.post(`${API_URL}/users/settings/notifications`, settings, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const setLanguage = createAsyncThunk<
  any,
  string,
  { rejectValue: { msg: string } }
>('user/setLanguage', async (languageCode, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.post(`${API_URL}/users/settings/language`, { language: languageCode }, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const fetchTransactionHistory = createAsyncThunk<
  any[],
  void,
  { rejectValue: { msg: string } }
>('user/fetchTransactionHistory', async (_, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.get(`${API_URL}/payment/history`, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

const userSlice = createSlice({
  name: 'user',
  initialState: {
    token: null,
    profile: null,
    profiles: [],
    conversations: [],
    blockedUsers: [],
    transactions: [],
    status: 'idle',
    error: null as string | null,
  },
  reducers: {
    logout: (state) => {
      state.token = null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(loginUser.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(loginUser.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.token = action.payload.token;
      })
      .addCase(loginUser.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Login failed';
      })
      .addCase(registerUser.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(registerUser.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.token = action.payload.token;
      })
      .addCase(registerUser.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Registration failed';
      })
      .addCase(createProfile.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(createProfile.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.profile = action.payload;
      })
      .addCase(createProfile.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Profile creation failed';
      })
      .addCase(updateUserProfile.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(updateUserProfile.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.profile = action.payload;
      })
      .addCase(updateUserProfile.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Profile update failed';
      })
      .addCase(fetchProfiles.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchProfiles.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.profiles = action.payload;
      })
      .addCase(fetchProfiles.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Fetching profiles failed';
      })
      .addCase(fetchConversations.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchConversations.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.conversations = action.payload;
      })
      .addCase(fetchConversations.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Fetching conversations failed';
      })
      .addCase(fetchUserProfile.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchUserProfile.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.profile = action.payload;
      })
      .addCase(fetchUserProfile.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Fetching user profile failed';
      })
      .addCase(likeUserProfile.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(likeUserProfile.fulfilled, (state, action) => {
        state.status = 'succeeded';
      })
      .addCase(likeUserProfile.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Liking user profile failed';
      })
      .addCase(fetchCurrentUserProfile.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchCurrentUserProfile.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.profile = action.payload;
      })
      .addCase(fetchCurrentUserProfile.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Fetching current user profile failed';
      })
      .addCase(subscribeUser.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(subscribeUser.fulfilled, (state, action) => {
        state.status = 'succeeded';
        // Update user profile to reflect premium status
        if (state.profile) {
          state.profile.isPremium = true;
        }
      })
      .addCase(subscribeUser.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Subscription failed';
      })
      .addCase(purchaseBoost.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(purchaseBoost.fulfilled, (state, action) => {
        state.status = 'succeeded';
        // Update user profile to reflect boost status if needed
        // This could involve setting a boost expiration timestamp
      })
      .addCase(purchaseBoost.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Boost purchase failed';
      })
      .addCase(fetchBlockedUsers.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchBlockedUsers.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.blockedUsers = action.payload;
      })
      .addCase(fetchBlockedUsers.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Failed to fetch blocked users';
      })
      .addCase(unblockUser.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(unblockUser.fulfilled, (state, action) => {
        state.status = 'succeeded';
        // Remove the unblocked user from the blocked users list
        if (state.blockedUsers) {
          state.blockedUsers = state.blockedUsers.filter((user: any) => user.id !== action.meta.arg);
        }
      })
      .addCase(unblockUser.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Failed to unblock user';
      })
      .addCase(updateNotificationSetting.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(updateNotificationSetting.fulfilled, (state, action) => {
        state.status = 'succeeded';
        // Update the user's notification settings
        if (state.profile) {
          state.profile.notificationSettings = {
            ...state.profile.notificationSettings,
            ...action.meta.arg
          };
        }
      })
      .addCase(updateNotificationSetting.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Failed to update notification settings';
      })
      .addCase(setLanguage.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(setLanguage.fulfilled, (state, action) => {
        state.status = 'succeeded';
        // Update the user's language preference
        if (state.profile) {
          state.profile.language = action.meta.arg;
        }
      })
      .addCase(setLanguage.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Failed to set language';
      })
      .addCase(fetchTransactionHistory.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchTransactionHistory.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.transactions = action.payload;
      })
      .addCase(fetchTransactionHistory.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Failed to fetch transaction history';
      });
  },
});

export const { logout } = userSlice.actions;
export const selectUserRegistrationStatus = (state: RootState) => state.user.status;
export const selectSubscriptionStatus = (state: RootState) => state.user.profile?.isPremium || false;
export const selectBoostPurchaseStatus = (state: RootState) => state.user.status;
export const selectBlockedUsers = (state: RootState) => state.user.blockedUsers || [];
export const selectNotificationSettings = (state: RootState) => state.user.profile?.notificationSettings || {};
export const selectCurrentLanguage = (state: RootState) => state.user.profile?.language || 'en';
export const selectAvailableLanguages = () => [
  { code: 'en', name: 'English' },
  { code: 'th', name: 'Thai' }
];
export const selectTransactionHistory = (state: RootState) => state.user.transactions || [];
export default userSlice.reducer;
