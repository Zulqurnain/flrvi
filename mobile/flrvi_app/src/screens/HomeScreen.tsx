import React, { useEffect } from 'react';
import { View, StyleSheet, FlatList, ActivityIndicator } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { fetchProfiles } from '../store/userSlice';
import { AppDispatch, RootState } from '../store/store';
import ProfileCard from '../components/ProfileCard';
import Header from '../components/Header';

const HomeScreen = ({ navigation }) => {
  const dispatch = useDispatch<AppDispatch>();
  const { profiles, status } = useSelector((state: RootState) => state.user);

  useEffect(() => {
    if (status === 'idle') {
      dispatch(fetchProfiles());
    }
  }, [status, dispatch]);

  const handleEndReached = () => {
    if (status !== 'loading') {
      // In a real app, you would dispatch fetchMoreProfiles here
    }
  };

  return (
    <View style={styles.container}>
      <Header
        title="Discover"
        leftAction={{
          icon: 'menu',
          action: () => navigation.openDrawer(),
        }}
        rightAction={{
          icon: 'filter',
          action: () => navigation.navigate('FilterModal'),
        }}
      />
      {status === 'loading' && profiles.length === 0 ? (
        <ActivityIndicator size="large" color="#E91E63" />
      ) : (
        <FlatList
          data={profiles}
          renderItem={({ item }) => <ProfileCard userProfile={item} onPress={() => navigation.navigate('ProfileScreen', { userId: item._id })} />}
          keyExtractor={(item) => item._id}
          numColumns={2}
          onEndReached={handleEndReached}
          onEndReachedThreshold={0.5}
          ListFooterComponent={status === 'loading' ? <ActivityIndicator /> : null}
        />
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#F5F5F5',
  },
});

export default HomeScreen;
