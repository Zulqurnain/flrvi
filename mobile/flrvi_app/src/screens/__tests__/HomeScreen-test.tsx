import React from 'react';
import renderer from 'react-test-renderer';
import HomeScreen from '../HomeScreen';
import { Provider } from 'react-redux';
import { store } from '../../store/store';

it('renders correctly', () => {
  const navigation = { navigate: jest.fn(), openDrawer: jest.fn() };
  const tree = renderer.create(
    <Provider store={store}>
      <HomeScreen navigation={navigation} />
    </Provider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
