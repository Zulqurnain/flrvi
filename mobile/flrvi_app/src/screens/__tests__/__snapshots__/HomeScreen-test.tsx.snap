// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#F5F5F5",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "elevation": 5,
        "flexDirection": "row",
        "height": 60,
        "justifyContent": "space-between",
        "padding": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
      }
    }
  >
    <View
      style={
        {
          "width": 50,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <
          color="black"
          name="menu"
          size={24}
        />
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        Discover
      </Text>
    </View>
    <View
      style={
        {
          "width": 50,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <
          color="black"
          name="filter"
          size={24}
        />
      </View>
    </View>
  </View>
  <RCTScrollView
    ListFooterComponent={null}
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
</View>
`;
